cumulusci/tasks/robotframework/robotperflistener.py

```diff
@@ -0,0 +1,149 @@
+        totals.update(
+            {
+                name: metric.value
+                for name, metric in self._current_test.custom_metrics.items()
+            }
+        )
+
+        test_result = {"totals": dict(sorted(totals.items()))}
+
+        if self.verbosity >= 1:
+            test_result["traces"] = self._current_test.api_metrics
+
+        self._current_suite[name] = test_result
+        self._current_test = None
+
+    def close(self):
+        filename = os.path.join(self.outputdir, "perf.json")
+        with open(filename, "w") as f:
+            json.dump(self._metrics, f, indent=2)
+
+        # cleanup so we won't be called again and also remove GC loop
+        self.task.robot_perf_listener = None
+        self.task = None
+
+    def output_file(self, path):
+        self.outputdir = os.path.dirname(path)
+
+    def report(self, metrics):
+        """Extension to the Listener API for reporting custom metrics."""
+        if self._current_test:  # sometimes we are called in suite teardown
+            self._current_test.api_metrics.append(metrics)
+
+    def create_aggregate_metric(self, name, aggregation):
+        """Extension to the Listener API for creating new aggregate metrics (.e.g. averages, sums)."""
+        self._current_test.custom_metrics[name] = Aggregations[aggregation]()
+
+    def store_metric_value(self, name, value):
+        """Extension to the Listener API for adding new values to a metric."""
+        self._current_test.custom_metrics[name].reduce(value)
+
+    def create_duration_metric(self, name):
+        """Extension to the Listener API for creating new durations."""
+        self._current_test.custom_metrics[name] = DurationAggregator()
+
+    def end_duration_metric(self, name):
+        """Extension to the Listener API for finishing durations."""
+        metric_obj = self._current_test.custom_metrics[name]
+        metric_obj.finish()
+
+
+class AverageAggregator:
+    """Takes a stream of numbers delivered to the "reduce" function and averages them"""
+
+    def __init__(self):
+        self.value = 0
+        self.count = 0
+
+    def reduce(self, value):
+        self.value = (self.value * self.count + float(value)) / (self.count + 1)
+        self.count += 1
+
+
+class TotalAggregator:
+    """Takes a stream of numbers delivered to the "reduce" function and sums them"""
+
+    def __init__(self):
+        self.value = 0
+
+    def reduce(self, value):
+        self.value += value
+
+
+class DurationAggregator:
+    """Figures out how long it takes to run some Robot code"""
+
+    def __init__(self):
+        self.start_time = time.time()
+        self.duration = -1
+
+    def reduce(self, value):
+        pass
+
+    def finish(self):
+        self.duration = time.time() - self.start_time
+
+    @property
+    def value(self):
+        return self.duration
+
+
+Aggregations = {
+    "duration": DurationAggregator,
+    "average": AverageAggregator,
+    "sum": TotalAggregator,
+}
```

Comment:
I like how this is structured but it's usually discouraged to name a Python identifier the same thing as as a builtin (reduce). I also always have found the programming term "reduce" to be pretty jargony. How about something like `aggregate` or `process`

Reply:
Sure, Process is fine. I did struggle with that name.

Reply:
56f2557da12f9fefd02d4421308ab58c022e3651